superset/datasets/commands/update.py

```diff
@@ -0,0 +1,83 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from flask_appbuilder.security.sqla.models import User
+from marshmallow import UnmarshalResult, ValidationError
+
+from superset.commands.base import BaseCommand, CommandValidateReturn
+from superset.datasets.commands.base import populate_owners
+from superset.datasets.commands.exceptions import (
+    DatabaseChangeValidationError,
+    DatasetExistsValidationError,
+    DatasetInvalidError,
+    DatasetNotFoundError,
+    DatasetUpdateFailedError,
+)
+from superset.datasets.dao import DatasetDAO
+from superset.views.base import check_ownership
+
+
+class UpdateDatasetCommand(BaseCommand):
+    def __init__(self, user: User, model_id: int, unmarshal: UnmarshalResult):
+        self._actor = user
+        self._model_id = model_id
+        self._properties = unmarshal.data.copy()
+        self._errors = unmarshal.errors
+        self._model = None
+
+    def run(self):
+        self._model = DatasetDAO.find_by_id(self._model_id)
+        if not self._model:
+            raise DatasetNotFoundError()
+        valid, exceptions = self.validate()
+        if not valid:
+            for exception in exceptions:
+                self._errors.update(exception.normalized_messages())
+            raise DatasetInvalidError()
+        dataset = DatasetDAO.update(self._model, self._properties)
+
+        if not dataset:
+            raise DatasetUpdateFailedError()
+        return dataset
+
+    def validate(self) -> CommandValidateReturn:
+        is_valid, exceptions = super().validate()
+        if not self._model:
```

Comment:
This is an interesting special case. We're checking for model presence in the `run` method of the command, and again here. I'd say the model lookup should either move inside of the `validate` call and raise, or this should return a meaningful error message, or we should keep the check in `run` and remove it from here. I'm not sure which one, but returning `False` with an empty array does not give the caller any meaningful information.

Reply:
Good point. Moved it to `validate` since I think it was a mistake to check it on `run` and making `validate` depend on it.